File: src/FSharpAux/Dictionary.fs
namespace FSharpAux

/// .Net Dictionary
module Dictionary = 
    
    open System
    open System.Collections.Generic

    /// <summary>Views the collection as an enumerable sequence of pairs.
    /// The sequence will be ordered by the keys of the dictionary.</summary>
    /// <param name="table">The input dictionary.</param>
    /// <returns>The sequence of key/value pairs.</returns>   
    let toSeq d = d |> Seq.map (fun (KeyValue(k,v)) -> (k,v))

    /// <summary>Returns a new dictionary made from the given bindings.</summary>
    /// <param name="elements">The input array of key/value pairs.</param>
    /// <returns>The resulting dictionary.</returns>  
    let toArray (d:IDictionary<_,_>) = d |> toSeq |> Seq.toArray
    
    /// <summary>Returns a new dictionary made from the given bindings.</summary>
    /// <param name="elements">The input list of key/value pairs.</param>
    /// <returns>The resulting dictionary.</returns>  
    let toList (d:IDictionary<_,_>) = d |> toSeq |> Seq.toList

    /// <summary>Builds a dictionary that contains the bindings of the given dictionary.</summary>
    /// <param name="elements">The input map.</param>
    /// <returns>The resulting dictionary.</returns> 
    let ofMap (m:Map<'k,'v>) = new Dictionary<'k,'v>(m) :> IDictionary<'k,'v>

    /// <summary>Builds a dictionary that contains the bindings of the given list.</summary>
    /// <param name="elements">The input list of key/value pairs.</param>
    /// <returns>The resulting dictionary.</returns> 
    let ofList (l:('k * 'v) list) = new Dictionary<'k,'v>(l |> Map.ofList) :> IDictionary<'k,'v>
    
    /// <summary>Builds a dictionary that contains the bindings of the given IEnumerable.</summary>
    /// <param name="elements">The input sequence of key/value pairs.</param>
    /// <returns>The resulting dictionary.</returns> 
    let ofSeq (s:('k * 'v) seq) = new Dictionary<'k,'v>(s |> Map.ofSeq) :> IDictionary<'k,'v>
    
    /// <summary>Builds a dictionary that contains the bindings of the given array.</summary>
    /// <param name="elements">The input array of key/value pairs.</param>
    /// <returns>The resulting dictionary.</returns> 
    let ofArray (a:('k * 'v) []) = new Dictionary<'k,'v>(a |> Map.ofArray) :> IDictionary<'k,'v>

        
    /// <summary>Returns the dictionary with the binding added to the given dictionary.
    /// If a binding with the given key already exists in the input dictionary, System.ArgumentException is thrown.</summary>
    /// <param name="key">The input key.</param>
    /// <returns>The dictionary with change in place.</returns>
    let addInPlace key value (table:IDictionary<_,_>) =
        table.Add(key,value)
        table

    /// <summary>Returns the dictionary with the binding added to the given dictionary.
    /// If a binding with the given key already exists in the input dictionary, the existing binding is replaced by the new binding in the result dictionary.</summary>
    /// <param name="key">The input key.</param>
    /// <returns>The dictionary with change in place.</returns>
    let addOrUpdateInPlace key value (table:IDictionary<_,_>) =
        match table.ContainsKey(key) with
        | true  -> table.[key] <- value
        | false -> table.Add(key,value)
        table

    /// <summary>Returns the dictionary with the binding added to the given dictionary.
    /// If a binding with the given key already exists in the input dictionary, the existing binding is replaced by the result of the given function.</summary>
    /// <param name="f">The function to aggregate old value and new value.</param>
    /// <param name="key">The input key.</param>    
    /// <returns>The dictionary with change in place.</returns>
    let addOrUpdateInPlaceBy f key value (table:IDictionary<_,_>) =
        match table.ContainsKey(key) with
        | true  -> 
            let value' = table.[key]
            table.[key] <- f value' value
        | false -> table.Add(key,value)
        table

    /// <summary>Returns true if there are no bindings in the dictionary.</summary>
    let isEmpty (table:IDictionary<_,_>) = table.Count < 1


//    /// <summary>The empty dictionary.</summary>
//    [<GeneralizableValueAttribute>]
//    [<CompiledName("Empty")>]
//    let empty<'Key,'T> : IDictionary<'Key,'T> when 'Key : comparison


    /// <summary>Tests if an element is in the domain of the dictionary.</summary>
    /// <param name="key">The input key.</param>
    /// <returns>True if the dictionary contains the given key.</returns>
    let containsKey key (table:IDictionary<_,_>) = table.ContainsKey(key)

    /// <summary>The number of bindings in the dictionary.</summary>
    let count (table:IDictionary<_,_>) = table.Count

    /// <summary>Lookup an element in the dictionary. Raise <c>KeyNotFoundException</c> if no binding
    /// exists in the dictionary.</summary>
    /// <param name="key">The input key.</param>
    /// <exception cref="System.Collections.Generic.KeyNotFoundException">Thrown when the key is not found.</exception>
    /// <returns>The value mapped to the key.</returns>
    let item key (table:IDictionary<_,_>) = table.[key]

    /// <summary>Removes an element from the domain of the dictionary. No exception is raised if the element is not present.</summary>
    /// <param name="key">The input key.</param>
    /// <returns>The resulting dictionary.</returns>
    let remove (key:'Key) (table:IDictionary<'Key,_>) = 
        table.Remove(key) |> ignore
        table

    /// <summary>Lookup an element in the dictionary, returning a <c>Some</c> value if the element is in the domain 
    /// of the dictionary and <c>None</c> if not.</summary>
    /// <param name="key">The input key.</param>
    /// <returns>The mapped value, or None if the key is not in the dictionary.</returns>
    let tryFind key (table:IDictionary<_,_>) =
        match table.ContainsKey(key) with
        | true -> Some table.[key]
        | false -> None


//    /// <summary>Folds over the bindings in the dictionary </summary>
//    /// <param name="folder">The function to update the state given the input key/value pairs.</param>
//    /// <param name="state">The initial state.</param>
//    /// <param name="table">The input dictionary.</param>
//    /// <returns>The final state value.</returns>
//    //[<CompiledName("Fold")>]
//    let fold<'Key,'T,'State> (f: 'State -> 'Key -> 'T -> 'State) (x:'State) (source : IDictionary<'Key,'T>)  = 
//        checkNonNull "source" source
//        use e = source.GetEnumerator() 
//        let f = OptimizedClosures.FSharpFunc<_,_,_,_>.Adapt(f)
//        let mutable state = x 
//        while e.MoveNext() do
//            state <- f.Invoke(state,e.Current.Key, e.Current.Value)
//        state


    // f is the function how to handel key conflicts
    /// Merge two Dictionaries
    let merge (a : IDictionary<'k,'v>) (b : IDictionary<'k,'v>) (f : 'k -> 'v -> 'v -> 'v) =        
        let dict =  new Dictionary<'k,'v>(a)
        for kv in b do           
            addOrUpdateInPlaceBy (f kv.Key) kv.Key kv.Value dict |> ignore
        dict :> IDictionary<'k,'v>

File: tests/FSharpAux.Mocha.Tests/fable/src/FSharpAux.Core/Dictionary.js
import { toList as toList_1, toArray as toArray_1, map } from "../../fable_modules/fable-library.4.0.6/Seq.js";
import { Dictionary } from "../../fable_modules/fable-library.4.0.6/MutableMap.js";
import { curry3, uncurry2, disposeSafe, getEnumerator, count as count_1, compare, structuralHash, equals } from "../../fable_modules/fable-library.4.0.6/Util.js";
import { ofArray as ofArray_1, ofSeq as ofSeq_1, ofList as ofList_1 } from "../../fable_modules/fable-library.4.0.6/Map.js";
import { getItemFromDict, addToDict } from "../../fable_modules/fable-library.4.0.6/MapUtil.js";
import { some } from "../../fable_modules/fable-library.4.0.6/Option.js";

export function toSeq(d) {
    return map((_arg) => {
        const activePatternResult = _arg;
        return [activePatternResult[0], activePatternResult[1]];
    }, d);
}

export function toArray(d) {
    return toArray_1(toSeq(d));
}

export function toList(d) {
    return toList_1(toSeq(d));
}

export function ofMap(m) {
    return new Dictionary(m, {
        Equals: equals,
        GetHashCode: structuralHash,
    });
}

export function ofList(l) {
    return new Dictionary(ofList_1(l, {
        Compare: compare,
    }), {
        Equals: equals,
        GetHashCode: structuralHash,
    });
}

export function ofSeq(s) {
    return new Dictionary(ofSeq_1(s, {
        Compare: compare,
    }), {
        Equals: equals,
        GetHashCode: structuralHash,
    });
}

export function ofArray(a) {
    return new Dictionary(ofArray_1(a, {
        Compare: compare,
    }), {
        Equals: equals,
        GetHashCode: structuralHash,
    });
}

export function addInPlace(key, value, table) {
    addToDict(table, key, value);
    return table;
}

export function addOrUpdateInPlace(key, value, table) {
    if (table.has(key)) {
        table.set(key, value);
    }
    else {
        addToDict(table, key, value);
    }
    return table;
}

export function addOrUpdateInPlaceBy(f, key, value, table) {
    if (table.has(key)) {
        const value$0027 = getItemFromDict(table, key);
        table.set(key, f(value$0027, value));
    }
    else {
        addToDict(table, key, value);
    }
    return table;
}

export function isEmpty(table) {
    return count_1(table) < 1;
}

export function containsKey(key, table) {
    return table.has(key);
}

export function count(table) {
    return count_1(table);
}

export function item(key, table) {
    return getItemFromDict(table, key);
}

export function remove(key, table) {
    table.delete(key);
    return table;
}

export function tryFind(key, table) {
    if (table.has(key)) {
        return some(getItemFromDict(table, key));
    }
    else {
        return void 0;
    }
}

export function merge(a, b, f) {
    const dict = new Dictionary(a, {
        Equals: equals,
        GetHashCode: structuralHash,
    });
    const enumerator = getEnumerator(b);
    try {
        while (enumerator["System.Collections.IEnumerator.MoveNext"]()) {
            const kv = enumerator["System.Collections.Generic.IEnumerator`1.get_Current"]();
            addOrUpdateInPlaceBy(uncurry2(curry3(f)(kv[0])), kv[0], kv[1], dict);
        }
    }
    finally {
        disposeSafe(enumerator);
    }
    return dict;
}

